lib/scholar/cluster/k_means.ex

```diff
@@ -0,0 +1,337 @@
+defmodule Scholar.Cluster.KMeans do
+  @moduledoc """
+  K-Means algorithm.
+  """
+  import Nx.Defn
+  import Scholar.Shared
+
+  @derive {Nx.Container, containers: [:num_iterations, :clusters, :inertia, :labels]}
+  defstruct [:num_iterations, :clusters, :inertia, :labels]
+
+  @doc """
+  Fits a K-Means model for sample inputs `x`
+
+
+  ## Options
+
+    * `:num_clusters` - The number of clusters to form as well as the number of centroids to generate. Required.
+
+    * `:max_iterations` - Maximum number of iterations of the k-means algorithm for a single run. Defaults to `300`.
+
+    * `:num_runs` - Number of time the k-means algorithm will be run with different centroid seeds.
+      The final results will be the best output of num_runs runs in terms of inertia. Defaults to `10`.
+
+    * `:tol` - Relative tolerance with regards to Frobenius norm of the difference in the cluster centers of two
+      consecutive iterations to declare convergence. Defaults to `1e-4`.
+
+    * `:random_state` - Determines random number generation for centroid initialization.
+      Use an int to make the randomness deterministic.
+      The argument `nil` is special and means that the seed is not set. Defaults to `nil`.
+
+    * `:weights` - The weights for each observation in x. If equals to `nil`, all observations
+      are assigned equal weight.
```

Comment:
It seems like this should be a tensor argument, as we convert it to a tensor anyway. Since it is one weight per each data point in `x`, the weights may be huge and we don't want to embed all of that as a constant into the computation.

One option is:

```elixir
defn fit(x, opts \\ []) do
  weights = Nx.broadcast(1, ...)
  fit(x, weights, opts)
end

defn fit(x, weights, opts) do
  ...
end
```

@polvalente @josevalim thoughts?

Reply:
FTR we can support tensors in options by having `fit/2` as a transform and let it call the `fit/3` defn. We will worry about this later.